server/src/lib/initializer.rs

```diff
@@ -237,111 +255,183 @@ impl<'a> MachineInitializer<'a> {
+                    let _ = self.inv.register_child(child, id).unwrap();
+                    backend.attach(vioblk.clone(), self.disp)?;
+                    chipset.device().pci_attach(bdf, vioblk);
+                }
+                StorageDeviceKind::Nvme => {
+                    let nvme = nvme::PciNvme::create(name.to_string(), be_info);
+                    let id =
+                        self.inv.register_instance(&nvme, bdf.to_string())?;
+                    let _ = self.inv.register_child(child, id).unwrap();
+                    backend.attach(nvme.clone(), self.disp)?;
+                    chipset.device().pci_attach(bdf, nvme);
+                }
+            };
+            if let Some((id, backend)) = crucible {
+                let prev = crucible_backends.insert(id, backend);
+                if prev.is_some() {
+                    return Err(Error::new(
+                        ErrorKind::InvalidInput,
+                        format!("multiple disks with id {}", id),
+                    ));
+                }
             }
-            _ => Err(std::io::Error::new(
-                std::io::ErrorKind::Other,
-                "Bad disk device!",
-            )),
         }
+        Ok(crucible_backends)
     }
 
-    pub fn initialize_in_memory_virtio_from_bytes(
+    pub fn initialize_network_devices(
         &self,
         chipset: &RegisteredChipset,
-        name: impl Into<String>,
-        bytes: Vec<u8>,
-        bdf: pci::Bdf,
-        read_only: bool,
     ) -> Result<(), Error> {
-        info!(self.log, "Creating in-memory disk from bytes");
-        let be =
-            propolis::block::InMemoryBackend::create(bytes, read_only, 512)?;
-
-        info!(self.log, "Creating ChildRegister");
-        let creg = ChildRegister::new(&be, Some(name.into()));
-
-        info!(self.log, "Calling initialize_virtio_block");
-        self.initialize_virtio_block(chipset, bdf, be, creg)
+        for (name, vnic_spec) in &self.spec.network_devices {
+            info!(self.log, "Creating vNIC {}", name);
+            let backend_spec = self
+                .spec
+                .network_backends
+                .get(&vnic_spec.backend_name)
+                .ok_or_else(|| {
+                    Error::new(
+                        ErrorKind::InvalidInput,
+                        format!(
+                            "Backend {} not found for vNIC {}",
+                            vnic_spec.backend_name, name
+                        ),
+                    )
+                })?;
+            let bdf: pci::Bdf = vnic_spec.pci_path.try_into().map_err(|e| {
+                Error::new(
+                    ErrorKind::InvalidInput,
+                    format!("Couldn't get PCI BDF for vNIC {}: {}", name, e),
+                )
+            })?;
+            let hdl = self.machine.get_hdl();
+            let viona = virtio::PciVirtioViona::new(
+                &backend_spec.vnic_name,
+                0x100,
```

Comment:
We should figure out how we want to encode the max ring size here, or frankly, any/all parameters of viona.  (this can be a follow-up)